src/CP/valueselection/learning/supervisedlearnedheuristic.jl

```diff
@@ -0,0 +1,143 @@
+    end
+end
+
+"""
+    (valueSelection::SupervisedLearnedHeuristic)(::InitializingPhase, model::CPModel, x::Union{Nothing, AbstractIntVar}, current_status::Union{Nothing, Symbol})
+
+Update the part of the SupervisedLearnedHeuristic which act like an RL environment, by initializing it with informations caracteritic of the new CPModel. Creates an
+initial observation with a false variable. A fixPoint! will be called before the SupervisedLearnedHeuristic takes its first decision.
+Finally, makes the agent call the process of the RL pre_episode_stage (basically making sure that the buffer is empty).
+"""
+function (valueSelection::SupervisedLearnedHeuristic)(::Type{InitializingPhase}, model::CPModel)
+    # create the environment
+    valueSelection.firstActionTaken = false
+    update_with_cpmodel!(valueSelection, model)
+    # FIXME get rid of this => prone to bugs
+    false_x = first(values(branchable_variables(model)))
+    env = get_observation!(valueSelection, model, false_x)
+    
+    eta = get_eta(valueSelection) #get the current eta_init 
+    if rand(valueSelection.rng) < eta
+        #the instance is solved using classic CP on a duplicated model
+        model_duplicate = deepcopy(model) 
+        strategy = DFSearch()
+        search!(model_duplicate, strategy, valueSelection.helpVariableHeuristic, valueSelection.helpValueHeuristic)
+        #the solution is added to the valueSelection.helpSolution field.
+        if !isnothing(model_duplicate.statistics.solutions)
+            solutions = model_duplicate.statistics.solutions[model_duplicate.statistics.solutions.!=nothing]
+            if length(solutions) >= 1
+                valueSelection.helpSolution = solutions[1]
+            end
+        end
+        reset_model!(model_duplicate)
+    else
+        valueSelection.helpSolution = nothing
+    end
+
+    # Reset the agent, useful for things like recurrent networks
+    Flux.reset!(valueSelection.agent)
+
+    if valueSelection.trainMode
+        valueSelection.agent(RL.PRE_EPISODE_STAGE, env)
+    end
+end
+
+"""
+    (valueSelection::SupervisedLearnedHeuristic)(::DecisionPhase, model::CPModel, x::Union{Nothing, AbstractIntVar}, current_status::Union{Nothing, Symbol})
+
+Observe, store useful informations in the buffer with agent(POST_ACT_STAGE, ...) and take a decision with the call to agent(PRE_ACT_STAGE).
+The metrics that aren't updated in the StepPhase, which are more related to variables domains, are updated here. Once updated, they can
+be used in the other set_reward! function as the reward of the last action will only be collected in the RL.POST_ACT_STAGE.
+"""
+function (valueSelection::SupervisedLearnedHeuristic)(PHASE::Type{DecisionPhase}, model::CPModel, x::Union{Nothing,AbstractIntVar})
+    # domain change metrics, set before reward is updated
+    set_metrics!(PHASE, valueSelection.search_metrics, model, x)
```

Comment:
Are we sure we want the metrics to be totally independent from the chosen value heuristic type?

Reply:
I'm not sure I understand, what would you consider here?